File: packages/sdk-go/internal/cli/help_test.go
package cli

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"
	"unicode/utf8"

	"github.com/leji-org/leji/packages/sdk-go/internal/commands/conformance"
	detectcmd "github.com/leji-org/leji/packages/sdk-go/internal/commands/detect"
	"github.com/leji-org/leji/packages/sdk-go/internal/detect"
	"github.com/leji-org/leji/packages/sdk-go/internal/ecosystem"
	"github.com/leji-org/leji/packages/sdk-go/internal/schemas"
	"github.com/leji-org/leji/packages/sdk-go/internal/writeplan"
)

// helpGolden reads one committed help golden: the byte oracle all three SDKs
// render against.
func helpGolden(t *testing.T, name string) string {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(repoRoot(t), "fixtures", "help-goldens", name))
	if err != nil {
		t.Fatalf("help golden %s: %v", name, err)
	}
	return string(data)
}

func goldenName(command string) string {
	return strings.ReplaceAll(command, " ", "-") + ".txt"
}

// hasDash reports U+2013 or U+2014: the house rule is that no line the CLI prints
// carries either.
func hasDash(s string) bool {
	return strings.ContainsAny(s, "–—")
}

// --- cli.json integrity: the grouping four consumers read ---------------------

func TestCliSpecGroupsAreWellFormed(t *testing.T) {
	spec, err := schemas.LoadCliSpec()
	if err != nil {
		t.Fatal(err)
	}
	ids := map[string]bool{}
	for _, g := range spec.Groups {
		if ids[g.ID] {
			t.Fatalf("duplicate group id %q", g.ID)
		}
		if g.Title == "" {
			t.Fatalf("group %q has no title", g.ID)
		}
		ids[g.ID] = true
	}
	names := map[string]*schemas.CliCommand{}
	for i := range spec.Commands {
		names[spec.Commands[i].Name] = &spec.Commands[i]
	}
	for _, c := range spec.Commands {
		if !ids[c.Group] {
			t.Fatalf("%s names group %q, which is not declared", c.Name, c.Group)
		}
		if c.AliasOf == "" {
			continue
		}
		primary, ok := names[c.AliasOf]
		if !ok {
			t.Fatalf("%s aliases %q, which is not a command", c.Name, c.AliasOf)
		}
		if primary.AliasOf != "" {
			t.Fatalf("%s aliases %q, which is itself an alias", c.Name, c.AliasOf)
		}
	}
}

// --- the wrapper --------------------------------------------------------------

func TestWrapCollapsesHangsAndKeepsLongTokensWhole(t *testing.T) {
	cases := []struct {
		text                           string
		width, indentFirst, indentRest int
		want                           []string
	}{
		{"  one   two  ", 20, 0, 0, []string{"one two"}},
		{"", 20, 0, 0, nil},
		{"alpha beta gamma delta", 16, 0, 3, []string{"alpha beta gamma", "   delta"}},
		{"alpha beta gamma delta", 16, 0, 8, []string{"alpha beta gamma", "        delta"}},
		// A token wider than the line takes a line of its own rather than being
		// split: a URL or a flag spelling stays copyable.
		{"see https://leji.org/cli/#mounts-update-pin now", 20, 0, 3,
			[]string{"see", "   https://leji.org/cli/#mounts-update-pin", "   now"}},
	}
	for _, c := range cases {
		if got := wrap(c.text, c.width, c.indentFirst, c.indentRest); !reflect.DeepEqual(got, c.want) {
			t.Fatalf("wrap(%q) = %q, want %q", c.text, got, c.want)
		}
	}
}

func TestTopLevelUsageGoesThroughTheWrapper(t *testing.T) {
	// No current command is long enough to wrap this line, so the contract is pinned on
	// a vector instead: every emitted field passes the wrapper, never just the ones the
	// data happens to overflow today.
	usage := "Usage: leji mounts update-pin <name> [--to <oid>] [--allow-non-fast-forward] " +
		"[--fetch] [--dry-run] [--root <dir>] [--json]"
	got := strings.Join(wrap(usage, 80, 0, 7), "\n") + "\n"
	if want := helpGolden(t, "wrap-long-usage.txt"); got != want {
		t.Fatalf("long usage vector\n got: %q\nwant: %q", got, want)
	}
	if !strings.Contains(BuildUsage(), "\nUsage: leji <command> [options]\n") {
		t.Fatal("top-level help does not carry the usage line")
	}
}

func TestOverlongLabelTakesItsOwnLine(t *testing.T) {
	label := "--allow-non-fast-forward-with-a-very-long-spelling <oid>"
	summary := "Permit a target that is not a descendant of the current pin, in the one " +
		"spelling long enough to outgrow its column."
	got := strings.Join(helpRow(label, 23, summary), "\n") + "\n"
	if want := helpGolden(t, "row-overlong-label.txt"); got != want {
		t.Fatalf("overlong label vector\n got: %q\nwant: %q", got, want)
	}
	// The clamp is what makes an overlong label reachable: past 27 characters the flag
	// outgrows its own column.
	if col := optionColumn([]schemas.CliOption{{Flags: label}}); col != 33 {
		t.Fatalf("optionColumn = %d, want 33", col)
	}
	// A label that exactly fills the column would leave no gap, so it takes the line too.
	want := []string{"   --exactly-here", "                 summary"}
	if got := helpRow("--exactly-here", 17, "summary"); !reflect.DeepEqual(got, want) {
		t.Fatalf("exact-fit row = %q, want %q", got, want)
	}
}

func TestRowPadsByCodePoints(t *testing.T) {
	// Two U+1F600 in the label: padding by UTF-16 units (or bytes) leaves the row short
	// and misaligns every summary in the block.
	label := "--emoji-\U0001F600\U0001F600 <value>"
	summary := "A flag carrying astral characters, so a column padded in UTF-16 units " +
		"misaligns this row by two."
	got := strings.Join(helpRow(label, 23, summary), "\n") + "\n"
	if want := helpGolden(t, "row-non-bmp.txt"); got != want {
		t.Fatalf("non-BMP row vector\n got: %q\nwant: %q", got, want)
	}
}

func TestEveryLabelClassResolvesABoundedColumn(t *testing.T) {
	opt := func(flags ...string) int {
		options := make([]schemas.CliOption, len(flags))
		for i, f := range flags {
			options[i] = schemas.CliOption{Flags: f}
		}
		return optionColumn(options)
	}
	name := func(names ...string) int {
		commands := make([]schemas.CliCommand, len(names))
		for i, n := range names {
			commands[i] = schemas.CliCommand{Name: n}
		}
		return nameColumn(commands)
	}
	code := func(codes ...string) int {
		exits := make([]schemas.CliExitCode, len(codes))
		for i, c := range codes {
			exits[i] = schemas.CliExitCode{Code: json.Number(c)}
		}
		return exitCodeColumn(exits)
	}
	cases := []struct {
		got, want int
		what      string
	}{
		{opt("--json"), 23, "option floor"},
		{opt("--a-flag-of-thirty-plus-characters <value>"), 33, "option ceiling"},
		{name("leji"), 15, "name floor"},
		{name("a-command-name-long-enough-to-outgrow-its-bounded-column"), 33, "name ceiling"},
		{code("0"), 6, "exit-code floor"},
		{code("0", "127"), 8, "exit-code widening"},
		// Code points, not bytes: an astral label sizes its column by what it prints.
		{opt("--emoji-\U0001F600\U0001F600 <value>"), 24, "astral option label"},
	}
	for _, c := range cases {
		if c.got != c.want {
			t.Fatalf("%s column = %d, want %d", c.what, c.got, c.want)
		}
	}
}

func TestBoundsHoldThroughTheRenderers(t *testing.T) {
	// Rendered, not just computed: a bound the column helper honors and the renderer
	// bypasses is exactly the defect this pins. The spec pushes every class past its
	// bound at once.
	raw, err := os.ReadFile(filepath.Join(repoRoot(t), "fixtures", "help-goldens", "bounds-spec.json"))
	if err != nil {
		t.Fatal(err)
	}
	var spec schemas.CliSpec
	if err := json.Unmarshal(raw, &spec); err != nil {
		t.Fatal(err)
	}
	want := strings.Replace(helpGolden(t, "bounds-usage.txt"), "{{version}}", schemas.SDKVersion, 1)
	if got := buildUsage(spec) + "\n"; got != want {
		t.Fatalf("synthetic top-level help\n got:\n%s\nwant:\n%s", got, want)
	}
	long := "a-command-name-long-enough-to-outgrow-its-bounded-column"
	help, ok := buildCommandHelp(long, spec)
	if !ok {
		t.Fatal("the synthetic command must render help")
	}
	if got, w := help+"\n", helpGolden(t, "bounds-command.txt"); got != w {
		t.Fatalf("synthetic command help\n got:\n%s\nwant:\n%s", got, w)
	}
}

func TestWrapMeasuresCodePointsNotBytes(t *testing.T) {
	// Documented in fixtures/README.md: four U+1F600, two spaces, three ASCII words,
	// width 20, first line indented 0 and continuations 3. Measuring the emoji run in
	// bytes (or UTF-16 units) breaks the line one word early.
	input := "\U0001F600\U0001F600\U0001F600\U0001F600  alphabet six666 tail"
	got := strings.Join(wrap(input, 20, 0, 3), "\n") + "\n"
	if want := helpGolden(t, "wrap-non-bmp.txt"); got != want {
		t.Fatalf("wrap vector\n got: %q\nwant: %q", got, want)
	}
}

// --- the goldens --------------------------------------------------------------

func TestHelpGoldens(t *testing.T) {
	spec, err := schemas.LoadCliSpec()
	if err != nil {
		t.Fatal(err)
	}
	want := strings.Replace(helpGolden(t, "usage.txt"), "{{version}}", schemas.SDKVersion, 1)
	if got := BuildUsage() + "\n"; got != want {
		t.Fatalf("leji --help does not match usage.txt\n got:\n%s\nwant:\n%s", got, want)
	}
	expected := []string{
		"usage.txt", "wrap-non-bmp.txt", "wrap-long-usage.txt", "row-overlong-label.txt",
		"row-non-bmp.txt", "bounds-spec.json", "bounds-usage.txt", "bounds-command.txt",
	}
	for _, c := range spec.Commands {
		help, ok := BuildCommandHelp(c.Name)
		if !ok {
			t.Fatalf("%s must render help", c.Name)
		}
		if got, w := help+"\n", helpGolden(t, goldenName(c.Name)); got != w {
			t.Fatalf("%s --help does not match %s\n got:\n%s\nwant:\n%s", c.Name, goldenName(c.Name), got, w)
		}
		expected = append(expected, goldenName(c.Name))
	}
	// And nothing committed is orphaned: every golden is one of the surfaces above.
	entries, err := os.ReadDir(filepath.Join(repoRoot(t), "fixtures", "help-goldens"))
	if err != nil {
		t.Fatal(err)
	}
	var found []string
	for _, e := range entries {
		found = append(found, e.Name())
	}
	sort.Strings(found)
	sort.Strings(expected)
	if !reflect.DeepEqual(found, expected) {
		t.Fatalf("help-goldens holds %v, want %v", found, expected)
	}
}

func TestCommandHelpListsOwnOptionsAndPointsAtGlobals(t *testing.T) {
	spec, err := schemas.LoadCliSpec()
	if err != nil {
		t.Fatal(err)
	}
	for _, c := range spec.Commands {
		help, _ := BuildCommandHelp(c.Name)
		if !strings.Contains(help, "\nGlobal options: see leji --help.\n") {
			t.Fatalf("%s help does not point at the globals", c.Name)
		}
		for _, g := range spec.GlobalOptions {
			if strings.Contains(help, "   "+g.Flags) {
				t.Fatalf("%s help repeats the global %s", c.Name, g.Flags)
			}
		}
		for _, o := range c.Options {
			if !strings.Contains(help, "   "+o.Flags) {
				t.Fatalf("%s help does not list %s", c.Name, o.Flags)
			}
		}
		// Examples are commands to copy, never prose: they are printed as authored,
		// so the width contract covers everything above them.
		prose, _, _ := strings.Cut(help, "\nExamples:\n")
		for _, line := range strings.Split(prose, "\n") {
			if utf8.RuneCountInString(line) > 80 {
				t.Fatalf("%s help line exceeds 80 code points: %q", c.Name, line)
			}
		}
	}
	for _, line := range strings.Split(BuildUsage(), "\n") {
		if utf8.RuneCountInString(line) > 80 {
			t.Fatalf("top-level help line exceeds 80 code points: %q", line)
		}
	}
}

// --- the em-dash house rule, checked on the bytes the CLI prints --------------

func TestHelpOutputCarriesNoDash(t *testing.T) {
	entries, err := os.ReadDir(filepath.Join(repoRoot(t), "fixtures", "help-goldens"))
	if err != nil {
		t.Fatal(err)
	}
	for _, e := range entries {
		if hasDash(helpGolden(t, e.Name())) {
			t.Fatalf("%s carries an em or en dash", e.Name())
		}
	}
}

func TestCliJSONCarriesNoDash(t *testing.T) {
	data, err := os.ReadFile(filepath.Join(repoRoot(t), "packages", "sdk-go", "internal", "assets", "cli.json"))
	if err != nil {
		t.Fatal(err)
	}
	if hasDash(string(data)) {
		t.Fatal("cli.json carries an em or en dash")
	}
}

func TestProseBranchesCarryNoDash(t *testing.T) {
	// detect's host lines: synthetic hosts, so the branch runs wherever the suite does.
	rendered := detectcmd.RenderDetect([]detect.DetectedHost{{
		ID:         "codex",
		Name:       "Codex CLI",
		Strength:   detect.Confirmed,
		OnPath:     true,
		InRepo:     true,
		UserConfig: false,
		Adapter:    "AGENTS.md",
	}}, ecosystem.Detect(t.TempDir()))
	if !strings.Contains(rendered, "Codex CLI: binary on PATH") {
		t.Fatalf("detect host line, got:\n%s", rendered)
	}
	if hasDash(rendered) {
		t.Fatalf("detect output carries a dash:\n%s", rendered)
	}

	// conformance --explain's blocker details, likewise: the detail branch needs a
	// blocker that carries one, which a passing layer does not produce.
	explain := conformance.RenderExplain(conformance.Result{
		ClaimedLevel:  "core",
		VerifiedLevel: "core",
		Items: []conformance.ChecklistItem{{
			ID:          "index-current",
			Level:       "indexed",
			Description: "a generated context index, current with the tree",
			Status:      conformance.Fail,
			Detail:      "the stored index is stale",
		}},
	})
	if !strings.Contains(explain, "- a generated context index, current with the tree: the stored index is stale") {
		t.Fatalf("explain detail line, got:\n%s", explain)
	}
	if hasDash(explain) {
		t.Fatalf("explain output carries a dash:\n%s", explain)
	}

	// The conformance checklist's own detail column, from a real run.
	example := filepath.Join(repoRoot(t), "examples", "monorepo")
	_, out, _ := captureRun(t, []string{"conformance", "--root", example})
	if !strings.Contains(out, "freshness horizons are declared and checked (report-only is acceptable): ") {
		t.Fatalf("checklist detail column, got:\n%s", out)
	}
	if hasDash(out) {
		t.Fatalf("conformance output carries a dash:\n%s", out)
	}

	// The write plan's read-only note is library data rather than a printed line, so
	// it is asserted where it is produced.
	plan := writeplan.Build(example, nil, []string{"README.md"}, nil)
	if plan[0].Note != "existing file, read-only input; Leji will not modify it" {
		t.Fatalf("write-plan note %q", plan[0].Note)
	}
}

// The unknown-command contract the help surface leans on: exit 2, the error, and
// the top-level usage, all on stderr.
func TestUnknownCommandPrintsUsageToStderr(t *testing.T) {
	code, _, errs := captureRun(t, []string{"frobnicate"})
	if code != 2 {
		t.Fatalf("unknown command exit %d, want 2", code)
	}
	if !strings.Contains(errs, `unknown command "frobnicate"`) {
		t.Fatalf("stderr %q does not name the unknown command", errs)
	}
	if !strings.Contains(errs, "Usage: leji") {
		t.Fatalf("stderr %q does not carry the top-level usage", errs)
	}
}

File: packages/sdk/test/help.test.ts
import { strict as assert } from 'node:assert';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { test } from 'node:test';
import { fileURLToPath } from 'node:url';
import {
   type ConformanceResult,
   SDK_VERSION,
   buildWritePlan,
   loadCliSpec,
   renderCommandHelp,
   renderDetect,
   renderExplain,
   renderUsage,
   run,
} from '../dist/index.js';
import { exitCodeColumn, helpRow, nameColumn, optionColumn, wrap } from '../dist/lib/text.js';

const pkgRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const repoRoot = path.resolve(pkgRoot, '..', '..');
const goldensDir = path.join(repoRoot, 'fixtures', 'help-goldens');
const golden = (name: string) => fs.readFileSync(path.join(goldensDir, name), 'utf8');
const goldenName = (command: string) => command.replace(/ /g, '-') + '.txt';
const exampleDir = path.join(repoRoot, 'examples', 'monorepo');

/** U+2013 and U+2014: the house rule is that no line the CLI prints carries either. */
const DASHES = /[–—]/;

async function stdoutOf<T>(fn: () => T | Promise<T>): Promise<string> {
   const chunks: string[] = [];
   const log = console.log;
   const err = console.error;
   console.log = (...a: unknown[]) => void chunks.push(a.map(String).join(' ') + '\n');
   console.error = () => {};
   try {
      await fn();
      return chunks.join('');
   } finally {
      console.log = log;
      console.error = err;
   }
}

// --- cli.json integrity: the grouping four consumers read ---------------------

test('cli.json: group ids are unique, every command is in one known group, aliases resolve', () => {
   const spec = loadCliSpec();
   const ids = spec.groups.map((g) => g.id);
   assert.deepEqual([...new Set(ids)], ids, 'group ids are unique');
   for (const g of spec.groups) assert.ok(g.title.length > 0, `${g.id} has a title`);
   const names = new Set(spec.commands.map((c) => c.name));
   for (const c of spec.commands) {
      assert.ok(typeof c.group === 'string' && ids.includes(c.group), `${c.name} names a declared group`);
      if (c.aliasOf === undefined) continue;
      assert.ok(names.has(c.aliasOf), `${c.name} aliases an existing command`);
      const primary = spec.commands.find((p) => p.name === c.aliasOf);
      assert.equal(primary!.aliasOf, undefined, `${c.name} aliases a primary, never another alias`);
   }
});

// --- the wrapper --------------------------------------------------------------

test('wrap: collapses whitespace, hangs continuations, and never breaks an oversized token', () => {
   assert.deepEqual(wrap('  one   two  ', 20, 0, 0), ['one two']);
   assert.deepEqual(wrap('', 20, 0, 0), []);
   assert.deepEqual(wrap('alpha beta gamma delta', 16, 0, 3), ['alpha beta gamma', '   delta']);
   // The continuation indent counts against the width, not only the first line.
   assert.deepEqual(wrap('alpha beta gamma delta', 16, 0, 8), ['alpha beta gamma', '        delta']);
   // A token wider than the line takes a line of its own rather than being split:
   // a URL or a flag spelling stays copyable.
   assert.deepEqual(wrap('see https://leji.org/cli/#mounts-update-pin now', 20, 0, 3), [
      'see',
      '   https://leji.org/cli/#mounts-update-pin',
      '   now',
   ]);
});

test('the top-level Usage line goes through the wrapper (shared long-usage vector)', () => {
   // No current command is long enough to wrap this line, so the contract is pinned on
   // a vector instead: every emitted field passes the wrapper, never just the ones the
   // data happens to overflow today.
   const usage =
      'Usage: leji mounts update-pin <name> [--to <oid>] [--allow-non-fast-forward] [--fetch] [--dry-run] [--root <dir>] [--json]';
   assert.equal(wrap(usage, 80, 0, 7).join('\n') + '\n', golden('wrap-long-usage.txt'));
   assert.match(renderUsage(), /\nUsage: leji <command> \[options\]\n/);
});

test('a label wider than its column takes the line alone (shared overlong-label vector)', () => {
   const label = '--allow-non-fast-forward-with-a-very-long-spelling <oid>';
   const summary =
      'Permit a target that is not a descendant of the current pin, in the one spelling long enough to outgrow its column.';
   assert.equal(helpRow(label, 23, summary).join('\n') + '\n', golden('row-overlong-label.txt'));
   // The clamp is what makes an overlong label reachable: past 27 characters the flag
   // outgrows its own column.
   assert.equal(optionColumn([label]), 33);
   // A label that exactly fills the column would leave no gap, so it takes the line too.
   assert.deepEqual(helpRow('--exactly-here', 17, 'summary'), ['   --exactly-here', '                 summary']);
});

test('a row pads by code points, not UTF-16 units (shared non-BMP row vector)', () => {
   // Two U+1F600 in the label: padding by UTF-16 units leaves the row two columns short
   // and misaligns every summary in the block.
   const label = '--emoji-\u{1F600}\u{1F600} <value>';
   const summary = 'A flag carrying astral characters, so a column padded in UTF-16 units misaligns this row by two.';
   assert.equal(helpRow(label, 23, summary).join('\n') + '\n', golden('row-non-bmp.txt'));
});

test('every dynamic label class resolves a bounded, code-point column', () => {
   assert.equal(optionColumn(['--json']), 23); // below the floor: [20, 30]
   assert.equal(optionColumn(['--a-flag-of-thirty-plus-characters <value>']), 33); // above the ceiling
   assert.equal(nameColumn(['leji']), 15); // below the floor: [12, 30]
   assert.equal(nameColumn(['a-command-name-long-enough-to-outgrow-its-bounded-column']), 33);
   assert.equal(exitCodeColumn(['0']), 6); // below the floor: [3, 8]
   assert.equal(exitCodeColumn(['0', '127']), 8);
   // Code points, not UTF-16 units: an astral label sizes its column by what it prints.
   assert.equal(optionColumn(['--emoji-\u{1F600}\u{1F600} <value>']), 24); // 18 code points, not 20 UTF-16 units
});

test('the bounds hold through the renderers, on the shared synthetic spec', () => {
   // Rendered, not just computed: a bound that the column helper honors and the renderer
   // bypasses is exactly the defect this pins. The spec pushes every class past its
   // bound at once.
   const spec = JSON.parse(fs.readFileSync(path.join(goldensDir, 'bounds-spec.json'), 'utf8'));
   assert.equal(renderUsage(spec) + '\n', golden('bounds-usage.txt').replace('{{version}}', SDK_VERSION));
   const long = 'a-command-name-long-enough-to-outgrow-its-bounded-column';
   assert.equal(renderCommandHelp(long, spec) + '\n', golden('bounds-command.txt'));
});

test('wrap: measures width in code points, not UTF-16 units (the shared non-BMP vector)', () => {
   // Documented in fixtures/README.md: four U+1F600, two spaces, three ASCII words,
   // width 20, first line indented 0 and continuations 3. Measuring the emoji run as
   // 8 UTF-16 units instead of 4 code points breaks the line one word early.
   const input = '\u{1F600}\u{1F600}\u{1F600}\u{1F600}  alphabet six666 tail';
   assert.equal(wrap(input, 20, 0, 3).join('\n') + '\n', golden('wrap-non-bmp.txt'));
});

// --- the goldens --------------------------------------------------------------

test('help goldens: every rendered help matches its committed bytes', () => {
   const spec = loadCliSpec();
   assert.equal(renderUsage() + '\n', golden('usage.txt').replace('{{version}}', SDK_VERSION));
   for (const c of spec.commands) {
      assert.equal(renderCommandHelp(c.name) + '\n', golden(goldenName(c.name)), c.name);
   }
   // And nothing committed is orphaned: every golden is one of the surfaces above.
   const expected = new Set([
      'usage.txt',
      'wrap-non-bmp.txt',
      'wrap-long-usage.txt',
      'row-overlong-label.txt',
      'row-non-bmp.txt',
      'bounds-spec.json',
      'bounds-usage.txt',
      'bounds-command.txt',
      ...spec.commands.map((c) => goldenName(c.name)),
   ]);
   assert.deepEqual(new Set(fs.readdirSync(goldensDir)), expected);
});

test("help goldens: per-command help lists the command's own options and points at the globals", () => {
   const spec = loadCliSpec();
   const globals = spec.globalOptions.map((o) => o.flags);
   // Examples are commands to copy, never prose: they are printed as authored, so
   // the width contract covers everything above them.
   const prose = (help: string) => help.split('\nExamples:\n')[0].split('\n');
   for (const c of spec.commands) {
      const help = renderCommandHelp(c.name)!;
      assert.match(help, /\nGlobal options: see leji --help\.\n/, c.name);
      for (const g of globals) assert.ok(!help.includes(`   ${g}`), `${c.name} does not repeat ${g}`);
      for (const o of c.options) assert.ok(help.includes(`   ${o.flags}`), `${c.name} lists ${o.flags}`);
      assert.ok(
         prose(help).every((l) => [...l].length <= 80),
         `${c.name} wraps at 80 code points`,
      );
   }
   assert.ok(
      renderUsage()
         .split('\n')
         .every((l) => [...l].length <= 80),
      'top-level help wraps at 80 code points',
   );
});

// --- the em-dash house rule, checked on the bytes the CLI prints --------------

test('help output carries no em or en dash', () => {
   for (const file of fs.readdirSync(goldensDir)) {
      assert.ok(!DASHES.test(golden(file)), `${file} carries an em or en dash`);
   }
});

test('cli.json carries no em or en dash', () => {
   const raw = fs.readFileSync(path.join(pkgRoot, 'cli.json'), 'utf8');
   assert.ok(!DASHES.test(raw));
});

test('the CLI prose branches carry no em or en dash', async () => {
   // detect's host lines: synthetic hosts, so the branch runs wherever the suite does.
   const detect = renderDetect({
      hosts: [
         {
            id: 'codex',
            name: 'Codex CLI',
            strength: 'confirmed',
            onPath: true,
            inRepo: true,
            userConfig: false,
            adapter: 'AGENTS.md',
         },
      ],
      ecosystem: { selected: null, all: [], reason: 'none' },
   });
   assert.match(detect, /Codex CLI: binary on PATH/);
   assert.ok(!DASHES.test(detect));

   // conformance --explain's blocker details, likewise: the detail branch needs a
   // blocker that carries one, which a passing layer does not produce.
   const result: ConformanceResult = {
      claimedLevel: 'core',
      verifiedLevel: 'core',
      processAttested: 0,
      items: [
         {
            id: 'index-current',
            level: 'indexed',
            description: 'a generated context index, current with the tree',
            status: 'fail',
            detail: 'the stored index is stale',
         },
      ],
      findings: [],
   };
   const explain = renderExplain(result);
   assert.match(explain, /- a generated context index, current with the tree: the stored index is stale/);
   assert.ok(!DASHES.test(explain));

   // The conformance checklist's own detail column, from a real run.
   const checklist = await stdoutOf(() => run(['conformance', '--root', exampleDir]));
   assert.match(checklist, /freshness horizons are declared and checked \(report-only is acceptable\): /);
   assert.ok(!DASHES.test(checklist));

   // The write plan's read-only note is library data rather than a printed line, so
   // it is asserted where it is produced.
   const plan = buildWritePlan(exampleDir, [], ['README.md']);
   assert.equal(plan[0].note, 'existing file, read-only input; Leji will not modify it');
   assert.ok(!DASHES.test(plan.map((e) => e.note ?? '').join('\n')));
});
